File: backend/app/api/v1/receiving.py
"""
收貨 API（CURSOR_INSTRUCTIONS.md）
現階段回傳與 inventory_lots 對齊的 mock；之後接上 SQLAlchemy 查詢即可。
"""

from typing import Any, Dict, List, Optional

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session

from app.core.warehouse.receiving import ReceivingService
from app.core.warehouse.putaway import PutAwayEngine
from app.schemas.receiving import IQCRequest, ReceiveRequest, ReceiveResponse, ScanResult
from app.api.deps import get_db, get_current_user, require_role
from app.config import settings
from app.core.printing.label_printer import LabelPrinter

router = APIRouter(tags=["receiving"])


@router.get("/list")
def get_receiving_list(
    po_number: Optional[str] = Query(None, description="採購單號篩選"),
    status: Optional[str] = Query(None, description="lot_status 篩選"),
    db: Session = Depends(get_db),
):
    """
    Query inventory_lots with lot_status in QC_HOLD/PENDING_RECEIVE (or iqc_result PENDING)
    Serialize to ReceivingItemApiRaw snake_case shape.
    """
    service = ReceivingService(db)
    items = service.list_pending(po_number=po_number, status=status)
    
     # Map ORM objects to dict for JSON response matching existing frontend expectations
    rows = []
    for item in items:
        rows.append({
             "lot_id": item.lot_id,
             "po_number": item.po_number,
             "vendor_name": item.vendor_name,
             "internal_sku": item.internal_sku,
             "internal_lot_number": item.internal_lot_number,
             "internal_barcode": item.internal_barcode,
             "vendor_pn": item.vendor_pn,
             "vendor_lot_code": item.vendor_lot_code,
             "vendor_date_code": item.vendor_date_code,
             "original_barcode": item.original_barcode,
             "description": item.description,
             "quantity_on_hand": item.quantity_on_hand,
             "unit": item.unit,
             "lot_status": item.lot_status,
             "receive_date": item.receive_date.isoformat() if item.receive_date else None,
             "location_code": item.location_code,
             "iqc_result": item.iqc_result,
             "iqc_date": item.iqc_date.isoformat() if item.iqc_date else None,
             "iqc_inspector": item.iqc_inspector,
             "quality_notes": item.quality_notes,
         })

    return {"items": rows, "total": len(rows)}


@router.get("/{lot_id}")
def get_receiving_detail(lot_id: int, db: Session = Depends(get_db)):
    service = ReceivingService(db)
    item = service.get_lot_by_id(lot_id)
    
    if not item:
        raise HTTPException(status_code=404, detail="Lot not found")

    return {
         "lot_id": item.lot_id,
         "po_number": item.po_number,
         "vendor_name": item.vendor_name,
         "internal_sku": item.internal_sku,
         "internal_lot_number": item.internal_lot_number,
         "internal_barcode": item.internal_barcode,
         "vendor_pn": item.vendor_pn,
         "vendor_lot_code": item.vendor_lot_code,
         "vendor_date_code": item.vendor_date_code,
         "original_barcode": item.original_barcode,
         "description": item.description,
         "quantity_on_hand": item.quantity_on_hand,
         "unit": item.unit,
         "lot_status": item.lot_status,
         "receive_date": item.receive_date.isoformat() if item.receive_date else None,
         "location_code": item.location_code,
         "iqc_result": item.iqc_result,
         "iqc_date": item.iqc_date.isoformat() if item.iqc_date else None,
         "iqc_inspector": item.iqc_inspector,
         "quality_notes": item.quality_notes,
     }


@router.post("/scan")
def scan_barcode(
    payload: Dict[str, Any], 
    db: Session = Depends(get_db)
):
    """
    Accept ScanBarcodeRequest {barcode, vendorId?}. Use BarcodeParser to parse and return ScanResult.
    Do not create any DB rows. Return 400 if barcode cannot be parsed.
    """
    barcode = payload.get("barcode") or payload.get("scannedBarcode")
    vendor_id = payload.get("vendorId") or payload.get("vendor_id")

    if not barcode:
        raise HTTPException(status_code=400, detail="Barcode is required")

    service = ReceivingService(db)
    try:
        result = service.scan_barcode(barcode, vendor_id)
         # Map internal ScanResult to API response shape
        return {
             "success": True,
             "parsed": {
                 "vendorPn": result.vendor_pn,
                 "qty": result.qty,
                 "lotCode": result.lot_code,
                 "dateCode": result.date_code,
             },
             "patternUsed": result.pattern_used,
         }
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.post("/receive")
def receive_item(
    request: ReceiveRequest,
    db: Session = Depends(get_db),
    current_user: dict = Depends(get_current_user),
):
    """
    Parse body into ReceiveRequest (po_number, barcode/scannedBarcode, vendor_id, qty).
    Call ReceivingService(db).process_receipt(...). Return ReceiveResponse.
    On PO mismatch let service raise and translate to HTTP 400 with detail message.
    """
    service = ReceivingService(db)

    try:
         # process_receipt already returns a ReceiveResponse (success/lotId/
         # internalLotNumber/internalBarcode/labelUrl) and raises HTTPException 400
         # on validation failure.
        return service.process_receipt(
            po_number=request.poNumber,
            scanned_barcode=request.scannedBarcode or "",
            vendor_id=request.vendorId,
            quantity=request.quantity,
            executed_by=current_user["username"],   # 以登入者為準,不信任 body
         )
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.post("/iqc")
def complete_iqc(
    request: IQCRequest,
    db: Session = Depends(get_db),
    current_user: dict = Depends(require_role("admin", "qc")),  # IQC 限 QC/管理員
):
    """
    Parse IQCRequest {lotId, result, inspector, notes?}.
    Call ReceivingService(db).complete_iqc(lot_id, result, inspector, notes).
    Return {success, status, suggestedLocation?} (suggestedLocation optionally from PutAwayEngine.suggest_location when result==PASS).
    """
    service = ReceivingService(db)

    try:
         # complete_iqc returns {success, status, suggestedLocation} and updates the
         # lot (status + iqc fields), suggesting a putaway location when result==PASS.
        return service.complete_iqc(
            lot_id=request.lotId,
            result=request.result,
            inspector=current_user["username"],     # 以登入者為準,不信任 body
            notes=request.notes
         )
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.post("/print-label")
def print_label(payload: Dict[str, Any], db: Session = Depends(get_db)):
    lot_id = payload.get("lot_id")
    
    if not lot_id:
        raise HTTPException(status_code=400, detail="lot_id is required")

    service = ReceivingService(db)
    item = service.get_lot_by_id(lot_id)
    
    if not item:
        raise HTTPException(status_code=404, detail="Lot not found")

    printer = LabelPrinter()
    zpl = printer.render(item)

    # 記錄已換標(強制換標流程的前置條件)
    service.mark_label_printed(lot_id)

    printed = False
    
    if settings.ZEBRA_PRINTER_IP:
        try:
            printer.send_to_printer(zpl, settings.ZEBRA_PRINTER_IP)
            printed = True
        except Exception:
            printed = False
            
    return {
        "success": True,
        "zpl": zpl,
        "printed": printed
    }

File: frontend/src/app/api/receiving.ts
import axios from 'axios';
import type { ReceivingItem, ReceivingItemApiRaw, ReceivingListResponse } from '../types/receiving';

// 一律走相對路徑(dev 由 Vite proxy、正式由 nginx 轉發),不依賴 VITE_API_URL
const API_BASE_URL = '';

function normalizeStatus(s: string): ReceivingItem['lotStatus'] {
  const upper = String(s || '').toUpperCase();
  if (upper === 'AVAILABLE') return 'AVAILABLE';
  if (upper === 'RESERVED') return 'RESERVED';
  if (upper === 'QC_HOLD' || upper === 'IQC') return 'QC_HOLD';
  if (upper === 'QUARANTINE') return 'QUARANTINE';
  if (upper === 'EXPIRED') return 'EXPIRED';
  if (upper === 'SHIPPED') return 'SHIPPED';
  // Optional alias for legacy fallback data
  if (upper === 'PENDING_RECEIVE' || upper === 'PENDING') return 'QC_HOLD';
  return 'AVAILABLE';
}

export function mapRawToReceivingItem(raw: ReceivingItemApiRaw): ReceivingItem {
  return {
    lotId: raw.lot_id,
    poNumber: raw.po_number ?? '',
    vendorName: raw.vendor_name ?? '',
    internalSku: raw.internal_sku ?? '',
    internalLotNumber: raw.internal_lot_number ?? '',
    internalBarcode: raw.internal_barcode ?? '',
    vendorPn: raw.vendor_pn ?? '—',
    vendorLotCode: raw.vendor_lot_code ?? '—',
    vendorDateCode: raw.vendor_date_code ?? undefined,
    originalBarcode: raw.original_barcode ?? '',
    description: raw.description ?? '',
    quantityOnHand: raw.quantity_on_hand ?? 0,
    unit: raw.unit ?? 'PCS',
    lotStatus: normalizeStatus(raw.lot_status),
    receiveDate: raw.receive_date ?? '',
    locationCode: raw.location_code ?? undefined,
    iqcResult: raw.iqc_result ? String(raw.iqc_result).toUpperCase() as ReceivingItem['iqcResult'] : undefined,
    iqcDate: raw.iqc_date ?? undefined,
    iqcInspector: raw.iqc_inspector ?? undefined,
    qualityNotes: raw.quality_notes ?? undefined,
   };
}

export interface ScanBarcodeRequest {
  barcode: string;
  vendorId?: number;
}

export interface ScanBarcodeResponse {
  success: boolean;
  parsed?: {
    vendorPn: string;
    qty: number;
    lotCode: string;
    dateCode?: string;
   };
  patternUsed?: string;
}

export interface ProcessReceiptRequest {
  poNumber: string;
  barcode: string;
  vendorId: number;
  qty: number;
}

export interface ProcessReceiptResponse {
  success: boolean;
  lotId: number;
  internalLotNumber: string;
  internalBarcode: string;
  labelUrl: string;
}

export interface CompleteIQCRequest {
  lotId: number;
  result: 'PASS' | 'FAIL';
  inspector: string;
  notes?: string;
}

export async function getReceivingList(params?: {
  poNumber?: string;
  status?: string;
  page?: number;
  pageSize?: number;
}): Promise<ReceivingListResponse> {
  const response = await axios.get<{ items: ReceivingItemApiRaw[]; total: number }>(`${API_BASE_URL}/api/v1/receiving/list`, {
    params: {
      po_number: params?.poNumber,
      status: params?.status,
      page: params?.page,
      page_size: params?.pageSize,
     },
   });
  const { items, total } = response.data;
  return {
    items: items.map(mapRawToReceivingItem),
    total: total ?? items.length,
   };
}

export async function getReceivingDetail(lotId: number): Promise<ReceivingItem> {
  const response = await axios.get<ReceivingItemApiRaw>(`${API_BASE_URL}/api/v1/receiving/${lotId}`);
  return mapRawToReceivingItem(response.data);
}

export async function scanBarcode(request: ScanBarcodeRequest): Promise<ScanBarcodeResponse> {
  const response = await axios.post<ScanBarcodeResponse>(`${API_BASE_URL}/api/v1/receiving/scan`, request);
  return response.data;
}

export async function processReceipt(request: ProcessReceiptRequest): Promise<ProcessReceiptResponse> {
  const response = await axios.post<ProcessReceiptResponse>(`${API_BASE_URL}/api/v1/receiving/receive`, {
    po_number: request.poNumber,
    barcode: request.barcode,
    vendor_id: request.vendorId,
    qty: request.qty,
  });
  return response.data;
}

export async function completeIQC(request: CompleteIQCRequest): Promise<{
  success: boolean;
  status: string;
  suggestedLocation?: string;
}> {
  const response = await axios.post(`${API_BASE_URL}/api/v1/receiving/iqc`, request);
  return response.data;
}

export async function printLabel(lotId: number): Promise<{
  success: boolean;
  zpl: string;
  printed: boolean;
}> {
  const response = await axios.post(`${API_BASE_URL}/api/v1/receiving/print-label`, { lot_id: lotId });
  return response.data;
}

// ==================== Purchase Order APIs ====================

export interface POItem {
  poNumber: string;
  vendorId: number;
  vendorName: string;
  poDate: string;
  status: string;
  lines: Array<{
    lineNumber: number;
    internalSku: string;
    vendorPn: string;
    orderedQty: number;
    receivedQty: number;
  }>;
}

export async function listPOs(): Promise<POItem[]> {
  const response = await axios.get<{ items?: POItem[] }>(`${API_BASE_URL}/api/v1/purchase-orders`);
  return Array.isArray(response.data) ? response.data : (response.data.items || []);
}

export interface ItemOption {
  internalSku: string;
  description: string;
}

export async function listItems(): Promise<ItemOption[]> {
  const response = await axios.get<{ items?: ItemOption[] }>(
    `${API_BASE_URL}/api/v1/purchase-orders/items`
  );
  return Array.isArray(response.data) ? response.data : (response.data.items || []);
}

export async function listOpenPOs(): Promise<string[]> {
  const response = await axios.get<{ items?: Array<{ poNumber: string; status: string }> }>(
    `${API_BASE_URL}/api/v1/purchase-orders/open-po-list`
  );
  const items = response.data.items || response.data;
  return Array.isArray(items) ? items.map((item: any) => item.poNumber) : [];
}

export interface Vendor {
  vendorId: number;
  vendorName: string;
}

export async function listVendors(): Promise<Vendor[]> {
  const response = await axios.get<{ items?: Vendor[] }>(`${API_BASE_URL}/api/v1/vendors`);
  return Array.isArray(response.data) ? response.data : (response.data.items || []);
}

export interface CreatePORequest {
  poNumber: string;
  vendorId: number;
  lines: Array<{
    internalSku: string;
    vendorPn: string;
    orderedQty: number;
  }>;
}

export interface CreatePOResponse {
  poNumber: string;
  vendorId: number;
  vendorName: string;
  poDate: string;
  status: string;
  lines: Array<{
    lineNumber: number;
    internalSku: string;
    vendorPn: string;
    orderedQty: number;
    receivedQty: number;
  }>;
}

export async function createPO(request: CreatePORequest): Promise<CreatePOResponse> {
  const response = await axios.post<CreatePOResponse>(`${API_BASE_URL}/api/v1/purchase-orders`, {
    poNumber: request.poNumber,
    vendorId: request.vendorId,
    lines: request.lines,
  });
  return response.data;
}
